runtime/lua/vim/lsp/client.lua

```diff
@@ -0,0 +1,473 @@
+
+    -- Data fields, to be used internally
+    __data__ = {},
+    attached_buf_list = {},
+    offset_encoding = 'utf16',
+    stdin = nil,
+    stdout = nil,
+    stderr = nil,
+    handle = nil,
+  }, client)
+
+  logger.info('Starting new client: ', server_name, cmd.execute_path, cmd.args)
+
+  return obj
+end
+
+client.start = function(self)
+  self.stdin = uv.new_pipe(false)
+  self.stdout = uv.new_pipe(false)
+  self.stderr = uv.new_pipe(false)
+
+  local function on_exit()
+    logger.info('filetype: '..self.filetype..', exit: '..self.cmd.execute_path)
+  end
+
+  local stdio = { self.stdin, self.stdout, self.stderr }
+  local opts = { args = self.cmd.args, stdio = stdio }
+  self.handle, self.pid = uv.spawn(self.cmd.execute_path, opts, on_exit)
+
+  uv.read_start(self.stdout, function (err, chunk)
+    if not err then
+      self:on_stdout(chunk)
+    end
+  end)
+
+  uv.read_start(self.stderr, function (err, chunk)
+    if err and chunk then
+      logger.error('stderr: '..err..', data: '..chunk)
+    end
+  end)
+
+  local bufs = vim.api.nvim_list_bufs()
+  for _, buf in ipairs(bufs) do
+    if vim.api.nvim_buf_get_option(buf, "ft") == self.filetype then
+      self:set_buf_change_handler(buf)
+    end
+  end
+
+  autocmd.register_text_document_autocmd(self.filetype, self.server_name)
+end
+
+client.stop = function(self)
+  if self._stopped then
+    return
+  end
+
+  autocmd.unregister_autocmd(self.filetype, self.server_name)
+
+  if (vim.api.nvim_call_function('exists', {'g:language_client_enable_shutdown_request'}) == 1) and
+    vim.api.nvim_get_var('language_client_log_level') ~= 0 then
+
+    vim.api.nvim_command("echo 'shutting down filetype: "..self.filetype.."server_name: "..self.server_name.."'")
+    self:request('shutdown', nil, function()end)
+  end
+
+  vim.api.nvim_command("echo 'exit filetype: "..self.filetype..", server_name: "..self.server_name.."'")
+  self:notify('exit', nil)
+
+  uv.shutdown(self.stdin, function()
+    uv.close(self.stdout)
+    uv.close(self.stderr)
+    uv.close(self.stdin)
+    uv.close(self.handle)
+  end)
+  uv.kill(self.pid, 'sigterm')
+
+  self._stopped = true
+end
+
+client.cmd_tostring = function(self)
+  local cmd = self.cmd
+  local cmd_str = cmd.execute_path
+  if cmd.args ~= nil then
+    for _, arg in pairs(cmd.args) do
+      cmd_str = cmd_str..' '..arg
+    end
+  end
+
+  return cmd_str
+end
+
+client.initialize = function(self)
+  local result = self:request_async('initialize', InitializeParams(self), function(_, data)
+    self:set_server_capabilities(data)
+    self:notify('initialized', {})
+    self:notify('textDocument/didOpen', DidOpenTextDocumentParams(self))
+    self:set_server_capabilities(data.capabilities)
+    return data.capabilities
+  end, nil)
+
+  return result
+end
+
+client.set_client_capabilities = function(self, capabilities)
+  self.client_capabilities = capabilities
+end
+
+client.set_server_capabilities = function(self, capabilities)
```

Comment:
this need to check if `offsetEncoding` was included (and change `client.offset_encoding` then, otherwise utf-16 should still be assumed).